File: frontend/src/lib/seriesCompact.test.ts
// seriesCompact.test.ts — v0.10.186 sözleşmesi (seriesCompact.ts başlığı; Go series_compact_test.go aynası).
import { describe, it, expect } from 'vitest';
import { expandCompactSeries, decodeBundle } from './seriesCompact';

describe('seriesCompact', () => {
  it('düzenli: t0 + i·step; düzensiz: t[]; tek/boş seri', () => {
    expect(expandCompactSeries({ groupKey: ['a'], t0: 100, step: 15, v: [1, 2.5, 3] })).toEqual({ groupKey: ['a'], points: [{ time: 100, value: 1 }, { time: 115, value: 2.5 }, { time: 130, value: 3 }] });
    expect(expandCompactSeries({ groupKey: ['g'], t: [0, 15, 45], v: [1, 2, 3] }).points.map(p => p.time)).toEqual([0, 15, 45]);
    expect(expandCompactSeries({ groupKey: ['b'], t0: 7, v: [0] })).toEqual({ groupKey: ['b'], points: [{ time: 7, value: 0 }] });
    expect(expandCompactSeries({ groupKey: ['c'], v: [] }).points).toEqual([]);
  });
  it('2^53 üstü unix ns: step tam saniye → t0 + i·step TAM (son nokta birebir)', () => {
    const t0 = 1_700_000_000_000_000_000, step = 15_000_000_000;
    const s = expandCompactSeries({ groupKey: ['x'], t0, step, v: new Array(240).fill(1) });
    expect(s.points[239].time).toBe(1_700_003_585_000_000_000); // 1.7e18 + 239·15e9, double'da tam
    expect(s.points[1].time - s.points[0].time).toBe(step);
  });
  it('decodeBundle: kodlu slot açılır (enc/cols düşer, öteki alanlar kalır); düz slot aynen; bilinmeyen enc → series SİLİNİR (self-fetch)', () => {
    const b = decodeBundle({
      p1: { enc: 'col', cols: [{ groupKey: ['x'], t0: 1, step: 1, v: [9] }], series: null, totalSeries: 5, tail: [{ time: 1, sum: 2, count: 1 }] },
      p2: { series: [{ groupKey: ['y'], points: [{ time: 1, value: 1 }] }] },
      p3: { enc: 'zzz', series: null },
    });
    expect(b.p1.series).toEqual([{ groupKey: ['x'], points: [{ time: 1, value: 9 }] }]);
    expect('enc' in b.p1).toBe(false);
    expect('cols' in b.p1).toBe(false);
    expect(b.p1.totalSeries).toBe(5);
    expect(b.p2.series?.[0].points[0].value).toBe(1);
    expect('series' in b.p3).toBe(false);
  });
});

File: internal/api/series_compact_test.go
package api

// series_compact_test.go — v0.10.186 sözleşmesi (series_compact.go başlığı).

import (
	"encoding/json"
	"reflect"
	"strings"
	"testing"

	"github.com/cilcenk/coremetry/internal/chstore"
)

func pts(t0, step int64, vals ...float64) []chstore.SpanMetricPoint {
	out := make([]chstore.SpanMetricPoint, len(vals))
	for i, v := range vals {
		out[i] = chstore.SpanMetricPoint{Time: t0 + int64(i)*step, Value: v}
	}
	return out
}

func TestCompactSeriesShapes(t *testing.T) {
	reg := chstore.SpanMetricSeries{GroupKey: []string{"a"}, Points: pts(1_700_000_000e9, 15e9, 1, 2.5, 3)}
	one := chstore.SpanMetricSeries{GroupKey: []string{"b"}, Points: pts(1_700_000_000e9, 15e9, 0)}
	empty := chstore.SpanMetricSeries{GroupKey: []string{"c"}}
	gap := chstore.SpanMetricSeries{GroupKey: []string{"g"}, Points: []chstore.SpanMetricPoint{{Time: 0, Value: 1}, {Time: 15e9, Value: 2}, {Time: 45e9, Value: 3}}}
	uniformGap := chstore.SpanMetricSeries{GroupKey: []string{"u"}, Points: pts(0, 300e9, 1, 2, 3)} // cron: tekdüze boşluk = düzenli
	cs := compactSeriesSet([]chstore.SpanMetricSeries{reg, one, empty, gap, uniformGap})
	if cs[0].T0 != 1_700_000_000e9 || cs[0].Step != 15e9 || cs[0].T != nil || cs[0].V[1] != 2.5 {
		t.Fatalf("düzenli: %+v", cs[0])
	}
	if cs[1].Step != 0 || len(cs[1].V) != 1 || cs[1].T != nil {
		t.Fatalf("tek nokta: %+v", cs[1])
	}
	if cs[2].Step != 0 || len(cs[2].V) != 0 || cs[2].T0 != 0 {
		t.Fatalf("boş: %+v", cs[2])
	}
	if cs[3].T == nil || cs[3].Step != 0 || !reflect.DeepEqual(cs[3].T, []int64{0, 15e9, 45e9}) {
		t.Fatalf("düzensiz açık zaman dizisi bekleniyor: %+v", cs[3])
	}
	if cs[4].Step != 300e9 || cs[4].T != nil {
		t.Fatalf("tekdüze boşluk düzenli sayılmalı: %+v", cs[4])
	}
	if _, ok := regularStep([]chstore.SpanMetricPoint{{Time: 5}, {Time: 5}}); ok {
		t.Fatal("sıfır adım düzenli sayıldı")
	}
	// yuvarlak-yolculuk: FE formülünün Go aynası girdiyi birebir geri verir
	for _, s := range []chstore.SpanMetricSeries{reg, one, gap, uniformGap} {
		if got := expandCompact(compactOne(s)); !reflect.DeepEqual(got.Points, s.Points) {
			t.Fatalf("round-trip bozuk: %+v → %+v", s.Points, got.Points)
		}
	}
	if got := expandCompact(compactOne(empty)); len(got.Points) != 0 {
		t.Fatalf("boş seri round-trip: %+v", got)
	}
}

func TestCompactSeriesBytesSmaller(t *testing.T) {
	vals := make([]float64, 240)
	for i := range vals {
		vals[i] = float64(i) * 1.234567
	}
	s := []chstore.SpanMetricSeries{{GroupKey: []string{"payments-orchestrator"}, Points: pts(1_700_000_000e9, 15e9, vals...)}}
	rowsJSON, _ := json.Marshal(s)
	colsJSON, _ := json.Marshal(compactSeriesSet(s))
	if len(colsJSON)*3 > len(rowsJSON) {
		t.Fatalf("sütunsal kodlama yeterince küçülmedi: rows=%d cols=%d", len(rowsJSON), len(colsJSON))
	}
	// kodlanmış slot: series:null + cols (FE null≠undefined kuralı: cols çözülür)
	slot := bundleSlot{Series: nil, Enc: seriesEncColumnar, Cols: compactSeriesSet(s)}
	b, _ := json.Marshal(slot)
	if !strings.Contains(string(b), `"series":null`) || !strings.Contains(string(b), `"cols":[`) || !strings.Contains(string(b), `"enc":"col"`) {
		t.Fatalf("kodlanmış slot şekli: %s", b[:120])
	}
}

// Kaynak kapısı: kodlama yalnız istemci opt-in'iyle (enc:"col") ve anahtar v3 (#2/#3).
func TestBundleEncodingGated(t *testing.T) {
	src := readSrc(t, "dashboards_data.go")
	for _, want := range []string{`dash-panel:v3`, `req.Enc == seriesEncColumnar`, `compactSeriesSet(slot.Series)`} {
		if !strings.Contains(src, want) {
			t.Fatalf("dashboards_data.go %q içermiyor", want)
		}
	}
}
